### PR TITLE
feat(cart): add dependency injectable reducers

### PR DESCRIPTION
**Pull request by griest024**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

`@daffodil/cart` provides no way for dependents to modify state reduction.

#1653

## What is the new behavior?
`@daffodil/cart` has mechanisms for providing reducers that run before or after the vanilla Daffodil cart reducers.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

### CODE REVIEW

File: libs/cart/state/src/reducers/token/post.token.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ * Provides reducers that run after the standard Daffodil cart reducers.
+ *
+ * ```ts
+ * providers: [
+ *   ...daffCartProvidePostReducers(
+ *     myReducer1,
+ *     myReducer2
+ *   )
+ * ]
```
Comment (lderrickable) [inline libs/cart/state/src/reducers/token/post.token.ts:34]: I think you need to close the "```ts" here and elsewhere

Comment (griest024) [reply libs/cart/state/src/reducers/token/post.token.ts:34]: fixed

**Review by damienwebdev — CHANGES_REQUESTED** [review]
Can you add a guide explaining when/why you would use this?

File: libs/cart/state/src/reducers/token/pre.token.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+import { DaffStatefulCartItem } from '../../models/public_api';
+import { DaffCartReducersState } from '../cart-reducers-state.interface';
+
+/**
+ * A token to hold the injectable pre reducers.
+ *
+ * Prefer using {@link daffCartProvidePreReducers}.
+ */
+export const DAFF_CART_PRE_REDUCERS = new InjectionToken<ActionReducer<DaffCartReducersState>[]>(
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/state/src/reducers/token/pre.token.ts]: changed

**Review by griest024 — COMMENTED** [review]
> Can you add a guide explaining when/why you would use this?

added

File: libs/cart/guides/state.md [hunk 3]
```diff
@@ -242,5 +242,71 @@ export class AppModule {}
...
+
+`@daffodil/cart/state` consumers may wish to modify the state for the `daffCart` feature in a way not explicitly provided by `@daffodil/cart/state`. A common example is adding payment info to the cart payment object in response to non-`@daffodil/cart/state` actions.
+
+#### Before vs. After
+
+Both before and after reducer slots are provided.
+
+Custom before reducers run before the native `@daffodil/cart/state` reducers. Before reducers can be used to contribute data to state with the understanding that the data may be overwritten by subsequent reducers.
+
+Custom after reducers run after the native `@daffodil/cart/state` reducers. After reducers can be used to fully control the final data in state. After reducers should take care to not malform state in a way that breaks `@daffodil/cart/state` features.
```
Comment (damienwebdev) [inline libs/cart/guides/state.md:259]: I think it's better to clearly state that as long as they meet the type constraint, this shouldn't be a problem.

File: libs/cart/guides/state.md [hunk 4]
```diff
@@ -242,5 +242,71 @@ export class AppModule {}
...
+
+### Purpose
+
+`@daffodil/cart/state` consumers may wish to modify the state for the `daffCart` feature in a way not explicitly provided by `@daffodil/cart/state`. A common example is adding payment info to the cart payment object in response to non-`@daffodil/cart/state` actions.
+
+#### Before vs. After
+
+Both before and after reducer slots are provided.
+
+Custom before reducers run before the native `@daffodil/cart/state` reducers. Before reducers can be used to contribute data to state with the understanding that the data may be overwritten by subsequent reducers.
```
Comment (damienwebdev) [inline libs/cart/guides/state.md:257]: It should also be noted that if a before reducers acts upon uninitialized state, that they must initialize state themselves.

**Review by damienwebdev — CHANGES_REQUESTED** [review]

File: libs/cart/guides/state.md (hunk 4 above)
Comment (griest024) [reply libs/cart/guides/state.md]: I removed before reducers entirely

File: libs/cart/guides/state.md (hunk 3 above)
Comment (griest024) [reply libs/cart/guides/state.md]: I added this to the above section

**Review by damienwebdev — CHANGES_REQUESTED** [review]

File: libs/cart/guides/state.md [hunk 5]
```diff
@@ -242,5 +242,63 @@ export class AppModule {}
...
+        },
+      };
+    default:
+      return state;
+  }
+}
+
+@NgModule({
+  providers: [
+    ...daffCartProvideAfterReducers(
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/guides/state.md]: fixed

**Comment by lderrickable** [comment]
lgtm

**Review by damienwebdev — APPROVED** [review]
lgtm

**Pull request merged**